pkg/handlers/generic/lifecycle/registry/utils/tls.go

```diff
@@ -0,0 +1,271 @@
+// Copyright 2025 Nutanix. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package utils
+
+import (
+	"cmp"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	_ "embed"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"net"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	handlersutils "github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pkg/handlers/utils"
+)
+
+const (
+	caCrtKey = "ca.crt"
+)
+
+var (
+	// Similar to CAPI, set the NotBefore to a few minutes in the past to account for clock skew.
+	// This cert is being generated on the management cluster, but used by a workload cluster.
+	defaultCertificateNotBeforeSkew = 5 * time.Minute
+	// Valid for 2 years to avoid expiring before the cluster is upgraded.
+	defaultCertificateDuration = 2 * 365 * 24 * time.Hour
+)
+
+type EnsureCertificateOpts struct {
+	// RemoteSecretKey is the name and namespace of the TLS secret to be created on the remote cluster.
+	RemoteSecretKey ctrlclient.ObjectKey
+
+	Spec CertificateSpec
+}
+
+type CertificateSpec struct {
+	// CommonName is the common name to be included in the certificate.
+	CommonName string
+	// DNSNames is a list of DNS names to be included in the certificate.
+	DNSNames []string
+	// IPAddresses is a list of IP addresses to be included in the certificate.
+	IPAddresses []string
+	// Duration is the duration for which the certificate is valid.
+	Duration time.Duration
+}
+
+// EnsureCASecretForCluster ensures that the registry addon CA secret exists for the given cluster.
+// It copies the ca.crt value from the global CA secret to a unique secret in the cluster's namespace.
+func EnsureCASecretForCluster(
+	ctx context.Context,
+	c ctrlclient.Client,
+	cluster *clusterv1.Cluster,
+) error {
+	globalTLSCertificateSecret, err := handlersutils.SecretForRegistryAddonRootCA(ctx, c)
+	if err != nil {
+		return err
+	}
+
+	clusterCASecret := buildClusterCASecret(globalTLSCertificateSecret, cluster)
+
+	// Copy the global CA certificate to a cluster CA secret.
+	err = handlersutils.EnsureSecretForLocalCluster(ctx, c, clusterCASecret, cluster)
```

Comment:
can you confirm the cluster CA will stay on the management cluster and regenerated every time on upgrade?

Reply:
Yes we expect the CA Secret for the cluster to exist, this handler will fail and mirror mutation handler will also fail.
Initially I copied the CA every time but removed that https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/1127/commits/9e22cd5ab45586013f765d27badd63e76f22901c#diff-d062d8df83a4dbee7f8632cbcb52dd12c887dad576ba3c5e9b4d1028e60953f7L91 since it was confusing why its in both places.

I think we can treat this the same like we do other cluster specific Secrets and just assume they are present.
Maybe using finalizers to prevent deletion but thats out of scope of this change IMO.